lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,333 @@
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+using namespace llvm;
+using namespace dynamatic;
+
+#define DEBUG_TYPE "cdg-analysis"
+
+void ControlDependenceAnalysis::identifyAllControlDeps(
+    mlir::func::FuncOp &funcOp) {
+  mlir::Region &funcReg = funcOp.getRegion();
+
+  // Initialize the all_control_deps_map by creating an entry for every block
+  // constituting the region
+  for (mlir::Block &block : funcReg.getBlocks()) {
+    llvm::SmallVector<mlir::Block *, 4> deps;
+    all_control_deps_map.insert(std::make_pair(&block, deps));
+  }
+
+  if (funcReg.hasOneBlock()) {
+    return; // functions made up of 1 basic block do not have any dependencies
+  }
+
+  // Get information about post-dominance
+  mlir::PostDominanceInfo postDomInfo;
+  // Get the post-dominance tree
+  llvm::DominatorTreeBase<mlir::Block, true> &postDomTree =
+      postDomInfo.getDomTree(&funcReg);
+
+  // Loop over the control flow edges connnecting the different blocks of this
+  // region
+  for (mlir::Block &block : funcReg.getBlocks()) {
+    for (mlir::Block *block_succ : block.getSuccessors()) {
+      if (!postDomInfo.properlyPostDominates(block_succ, &block)) {
+        mlir::Block *least_common_anc =
+            postDomInfo.findNearestCommonDominator(block_succ, &block);
+
+        if (least_common_anc == &block) {
+          // All nodes in the post-dominator tree on the path from the "block"
+          // to "block_succ" (including both "block" and "block_succ") should be
+          // control dependent on "block"
+
+          // easy case of block and block_succ
+          all_control_deps_map[&block].push_back(&block); // loop case
+          all_control_deps_map[block_succ].push_back(&block);
+
+          // traverse the tree to get all nodes between "least_common_anc" and
+          // "block_succ"
+          llvm::SmallVector<llvm::SmallVector<mlir::DominanceInfoNode *, 4>, 4>
+              traversed_nodes;
+          traversePostDomTree(&block, block_succ, &funcReg, &postDomTree,
+                              &traversed_nodes);
+
+          // prints the above traversal only in debug mode
+          printPostDomTreeTraversal(traversed_nodes);
+
+          for (size_t i = 0; i < traversed_nodes.size(); i++) {
+            for (size_t j = 0; j < traversed_nodes[i].size(); j++) {
+              // for every node in every path, add block to its control
+              // dependencies
+              mlir::Block *b = traversed_nodes[i][j]->getBlock();
+
+              if (b == &block || b == block_succ)
+                continue; // skip the start and the end nodes because we have
+                          // taken care of them above
+
+              if (std::find(all_control_deps_map[b].begin(),
+                            all_control_deps_map[b].end(),
+                            &block) == all_control_deps_map[b].end()) {
+                all_control_deps_map[b].push_back(&block);
+              }
+            }
+          }
```

Comment:
I think it is fine to merge the code in l62--l64 here to shorten the code
```c++
          // easy case of block and block_succ
          all_control_deps_map[&block].push_back(&block); // loop case
          all_control_deps_map[block_succ].push_back(&block);
```

---

I think unless needed, we should use range-based for loop instead of index-based (fewer codes), that is:

```c++
for (PathInDomTree path : allPathsFromLeastCommonAncToBlockSucc)
  for (DominanceInfoNode * domInfo : path) {
    Block *blockInPath = domInfo->getBlock();
    allControlDepsMap.insert(&block);
  }
```

I think you could (PLEASE CHECK) implement `allControlDepsMap` as a map from blocks to sets, so you don't need to explicitly check if there are duplicated elements in the set.

Reply:
Hi Jiahui, I do not think it would be a good idea to change it to a set because I added some additional code to allow us to modify the Block ptrs at any point. This is something I need in the cfToHandshake pass and I think modifying an element in a set is more complicated than modifying one in a vector.